### PR TITLE
chore: Move compression utilities into separate package

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does / why we need it**:

Compression tooling has been part of the `chunkenc` (chunk encoding)
package in the past for legacy reasons.

Since more components use this now, it's easier to keep it in a separate
package. This also eliminates the confusion around "encoding", since
this has been incorrectly used synonymously with "compression" in the past.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [x] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

Signed-off-by: Christian Haudum <christian.haudum@gmail.com>

### CODE REVIEW

File: pkg/compression/encoding.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
+package compression
+
+import (
+	"fmt"
+	"strings"
+)
+
+// Encoding is the identifier for a chunk encoding.
```
Comment (chaudum) [inline pkg/compression/encoding.go:8]: Not exclusively ....

File: pkg/compression/pool.go [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+		return &Snappy
+	case EncNone:
+		return &Noop
+	case EncFlate:
+		return &Flate
+	case EncZstd:
+		return &Zstd
+	default:
+		panic("unknown encoding")
+	}
```
Comment (rfratto) [inline pkg/compression/pool.go:70]: nit: None* of the pools are referenced directly outside of this package, so it's safe to unexport all of them and just leave `GetReaderPool`/`GetWriterPool` as the public API for pools. 

Personally I'm normally a fan of unexporting as much as possible as it reduces the surface where good API documentation is necessary. I don't feel too strongly about this at the moment, so I don't mind if this doesn't get changed. 

*: `compression.Gzip` is referenced a few times, but all references can be replaced with `compression.GetReaderPool(compression.EncGZIP)` or `compression.GetWriterPool(compression.EncGZIP)`.

File: pkg/compression/pool.go [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+}
+
+// GzipPool is a gun zip compression pool
```
Comment (rfratto) [inline pkg/compression/pool.go:73]: Is gun zip a typo?

**Review by rfratto — APPROVED** [review]

File: pkg/compression/encoding.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+type Encoding byte
+
+// The different available encodings.
+// Make sure to preserve the order, as these numeric values are written to the chunks!
```
Comment (rfratto) [inline pkg/compression/encoding.go:12]:
```suggestion
// Make sure to preserve the order, as the numeric values are serialized!
```

File: pkg/compression/encoding.go (hunk 1 above)
Comment (rfratto) [reply pkg/compression/encoding.go:8]:
```suggestion
// Encoding identifies an available compression type. 
```

**Review by rfratto — COMMENTED** [review]
(two small extra comments)

**Pull request merged**